### PR TITLE
Deprecation: State_vector not wavefunction, wave_function, or state.

### PR DESCRIPTION
**Pull request by dabacon**
Deprecates `wavefunction`, `wave_function` and `state` (when it is a state vector) to `state_vector`.  Did not change `state_rep` which is used when the state is supplied as either a state vector or computational basis state.   Added a new docs/dev markdown describing nomenclature preferences and linked from style.md.

In cirq.linalg:
- deprecates `subwavefunction` with `sub_state_vector`
- deprecates `wavefunction_partial_trace_as_mixture` with `partial_trace_of_state_vector_as_mixture`
- in `partial_trace_of_state_vector_as_mixture`, deprecates `state` parameter for `state_vector`
- in `sub_state_vector`, deprecates `state` parameter for `state_vector`

In cirq.qis:
- deprecates `validate_normalized_state` with `validate_normalized_state_vector`
- in `bloch_vector_from_state_vector` deprecates parameter `state` for `state_vector`.
- in `density_matrix_for_state_vector` deprecates parameter `state` for `state_vector`.
- in `dirac_notation` deprecates parameter `state` for `state_vector`.
- in `validate_qid_shape` deprecates parameter `state` for `state_vector`.

In cirq.sim:
- deprecates `final_wavefunction`  with `final_state_vector`
- deprecates `SimulatesIntermediateWaveFunction` with `SimulatesIntermediateStateVector`
- deprecates `WaveFunctionSimulatorState` with `StateVectorSimulatorState`
- deprecates `WaveFunctionStepResult` with `StateVectorStepResult`
- deprecates `WaveFunctionTrialResult` with `StateVectorTrialResult`
- deprecates `StateVectorTrialResult.final_state` with `StateVectorTrialResult.final_state_vector`
- renamed `state` parameter in `_StateAndBuffer` to `state_vector`. No deprecation because this is private.
- renamed wave_function.py state_vector.py
- updated deprecation deadline of `validate_normalized_state` in `wave_function.py` to `validate_normalized_state` which exists in `qis`.
- in `sample_state_vector` deprecates parameter `state` to `state_vector`.
-in `measure_state_vector` deprecates parameter `state` to `state_vector`.

In cirq.sim.clifford:
- deprecates `CliffordState.wave_function` to `CliffordState.state_vector`
- in `CliffordState.perform_measurement` deprecates parameter `collapse_wavefunction` with `collapse_state_vector`.
- deprecates`StabilizerStateChForm.wave_function` to `StabilizerStateChForm.state_vector`.

In cirq.circuits
- deprecates `Circuit.final_wavefunction` with `Circuit.final_state_vector`

In cirq.ops
- deprecates `PauliSum.expecation_from_wavefunction` for `PauliSum.expectation_from_state_vector`
- in `PauliSum.expectation_from_state_vector` deprecates parameter `state` for `state_vector`
- deprecates `PauliXtring.expecation_from_wavefunction` for `PauliString.expectation_from_state_vector`
- in `PauliString.expectation_from_state_vector` deprecates parameter `state` for `state_vector`

Also:

Rewrote tests in states_test to use names of states for bloch vector and other states, instead of the gates to produce them from |0>, and parameterized these tests for global phase.

Changed a FIXME to a TODO.

### CODE REVIEW

File: cirq/linalg/transformations.py [hunk 1]
```diff
@@ -317,53 +318,65 @@ def partial_trace(tensor: np.ndarray,
...
-        probability '1' if the factored state is pure, or else a mixture of the
-        default eigendecomposition of the mixed state's partial trace.
+        A single-component mixture in which the factored state_function has
+        probability '1' if the partially traced state is pure, or else a
+        mixture of the default eigendecomposition of the mixed state's
+        partial trace.
 
     Raises:
-        ValueError: if the input wavefunction is not an array of length
+        ValueError: if the input state_function is not an array of length
```
Comment (Strilanc) [inline cirq/linalg/transformations.py:356]: `state_function`?

File: cirq/linalg/transformations.py [hunk 2]
```diff
@@ -372,19 +385,32 @@ def wavefunction_partial_trace_as_mixture(
...
+def sub_state_vector(state_vector: np.ndarray,
+                     keep_indices: List[int],
+                     *,
+                     default: TDefault = RaiseValueErrorIfNotProvided,
+                     atol: Union[int, float] = 1e-8) -> np.ndarray:
+    r"""Attempts to factor a state_vector into two parts and return one of them.
+
+    The input state_vector must have shape ``(2,) * n`` or ``(2 ** n)`` where
+    `state_vector` is expressed over n qubits. The returned array will retain
+    the same type of shape as the input state_function, either ``(2 ** k)`` or
```
Comment (Strilanc) [inline cirq/linalg/transformations.py:410]: state_function

File: cirq/linalg/transformations.py [hunk 3]
```diff
@@ -317,53 +318,65 @@ def partial_trace(tensor: np.ndarray,
...
         atol: Union[int, float] = 1e-8) -> Tuple[Tuple[float, np.ndarray], ...]:
-    """Returns a mixture representing a wavefunction with only some qubits kept.
+    """Returns a mixture representing a state vector with only some qubits kept.
 
-    The input wavefunction must have shape `(2,) * n` or `(2 ** n)` where
-    `wavefunction` is expressed over n qubits. States in the output mixture will
-    retain the same type of shape as the input wavefunction, either `(2 ** k)`
+    The input state vector must have shape `(2,) * n` or `(2 ** n)` where
+    `state_vector` is expressed over n qubits. States in the output mixture will
+    retain the same type of shape as the input state_function, either `(2 ** k)`
```
Comment (Strilanc) [inline cirq/linalg/transformations.py:337]: state_function

File: cirq/linalg/transformations.py [hunk 4]
```diff
@@ -317,53 +318,65 @@ def partial_trace(tensor: np.ndarray,
...
+        state_vector: The state vector to take the partial trace over.
+        keep_indices: Which indices to take the partial trace of the
+            state_vector on.
         atol: The tolerance for determining that a factored state is pure.
 
     Returns:
-        A single-component mixture in which the factored wavefunction has
-        probability '1' if the factored state is pure, or else a mixture of the
-        default eigendecomposition of the mixed state's partial trace.
+        A single-component mixture in which the factored state_function has
```
Comment (Strilanc) [inline cirq/linalg/transformations.py:350]: state_function

File: cirq/linalg/transformations.py [hunk 5]
```diff
@@ -372,19 +385,32 @@ def wavefunction_partial_trace_as_mixture(
...
+                     atol: Union[int, float] = 1e-8) -> np.ndarray:
+    r"""Attempts to factor a state_vector into two parts and return one of them.
+
+    The input state_vector must have shape ``(2,) * n`` or ``(2 ** n)`` where
+    `state_vector` is expressed over n qubits. The returned array will retain
+    the same type of shape as the input state_function, either ``(2 ** k)`` or
     ``(2,) * k`` where k is the number of qubits kept.
 
-    If a wavefunction $|\psi\rangle$ defined on n qubits is an outer product
+    If a state_vector $|\psi\rangle$ defined on n qubits is an outer product
```
Comment (Strilanc) [inline cirq/linalg/transformations.py:413]: underscore doesn't really make sense in prose

File: cirq/qis/states.py [hunk 6]
```diff
@@ -93,66 +115,81 @@ def density_matrix_from_state_vector(
...
-    state = np.asarray(state).reshape(qid_shape)
+    state_vector = np.asarray(state_vector).reshape(shape)
 
     sum_inds = np.array(range(n_qubits))
     sum_inds[indices] += n_qubits
 
-    rho = np.einsum(state, list(range(n_qubits)), np.conj(state),
+    rho = np.einsum(state_vector, list(range(n_qubits)), np.conj(state_vector),
                     sum_inds.tolist(), indices + sum_inds[indices].tolist())
-    new_shape = np.prod([qid_shape[i] for i in indices], dtype=int)
```
Comment (Strilanc) [inline cirq/qis/states.py:129]: Unrelated rename?

**Comment by dabacon** [comment]
Couple of things just to make sure on:

Name: `subwavefunction` became `sub_state_vector` which is a bit awkward.
Name: `wavefunction_partial_trace_as_mixture` became `partial_trace_of_state_vector_as_mixture` which is super long, and changes the verb noun sense, but I think reads better.

Also trying to track down a doc string problem.

File: cirq/qis/states.py (hunk 6 above)
Comment (dabacon) [reply cirq/qis/states.py:129]: No, there is a bad no no aboe: qid_shape = validate_qid_shate(state, qid_state) could modify qid_shape.  Caught by linter, not sure why it wasn't picked up earlier.

File: cirq/circuits/circuit.py [hunk 7]
```diff
@@ -1494,7 +1495,7 @@ def unitary(self,
         result = _apply_unitary_circuit(self, state, qs, dtype)
         return result.reshape((side_len, side_len))
 
-    def final_wavefunction(
+    def final_state_vector(
             self,
             initial_state: 'cirq.STATE_VECTOR_LIKE' = 0,
```
Comment (dstrain115) [inline cirq/circuits/circuit.py:1500]: Should this be changed into initial_state_vector, or is this too generic?

File: cirq/qis/states_test.py [hunk 8]
```diff
@@ -36,46 +37,69 @@ def assert_valid_density_matrix(matrix, num_qubits=None, qid_shape=None):
...
-def test_bloch_vector_simple_H_zero():
-    sqrt = np.sqrt(0.5)
-    H_state = np.array([sqrt, sqrt])
+@pytest.mark.parametrize('global_phase', (1, 1j, np.exp(1j)))
+def test_bloch_vector_zero_state(global_phase):
+    zero_state = global_phase * np.array([1, 0])
 
-    bloch = cirq.bloch_vector_from_state_vector(H_state, 0)
-    desired_simple = np.array([1, 0, 0])
+    bloch = cirq.bloch_vector_from_state_vector(zero_state, 0)
```
Comment (dstrain115) [inline cirq/qis/states_test.py:44]: What's the deal with changing the test here?  Is there a good reason from going from H state to zero_state?

File: cirq/qis/states_test.py [hunk 9]
```diff
@@ -36,46 +37,69 @@ def assert_valid_density_matrix(matrix, num_qubits=None, qid_shape=None):
...
+    desired_simple = np.array([0, 0, 1])
     np.testing.assert_array_almost_equal(bloch, desired_simple)
 
 
-def test_bloch_vector_simple_XH_zero():
+@pytest.mark.parametrize('global_phase', (1, 1j, np.exp(1j)))
+def test_bloch_vector_one_state(global_phase):
+    zero_state = global_phase * np.array([0, 1])
```
Comment (dstrain115) [inline cirq/qis/states_test.py:51]: This says test_bloch_vector_**one_state** but then has zero_state.  Is this a typo?

File: cirq/sim/simulator.py [hunk 10]
```diff
@@ -203,7 +203,7 @@ def simulate(
         """Simulates the supplied Circuit.
 
         This method returns a result which allows access to the entire
-        wave function.
+        quantum computer's final state.
```
Comment (dstrain115) [inline cirq/sim/simulator.py:206]: quantum computer seems awkward here, since it is a simulator.  Should we say the simulator's final state?

File: cirq/sim/state_vector_simulator_test.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
+# Copyright 2019 The Cirq Developers
```
Comment (dstrain115) [inline cirq/sim/state_vector_simulator_test.py:1]: nit: 2020

File: docs/dev/nomenclature.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
+## Nomenclature
+
```
Comment (dstrain115) [inline docs/dev/nomenclature.md:2]: Can we write a short introduction here?

Something like, "This page describes naming rules and conventions that exist throughout cirq.  These rules are important so that we maintain a consistent interface that is easy to use...."

File: style.md [hunk 13]
```diff
@@ -69,3 +69,8 @@ Note that if you do this you will need to use the string version of the type,
 def my_func() -> 'module.that.causes.cycle.MyClass':
     pass
 ```
+
+#### Nomenclature
+
+For a list of terms we prefer for consistency see
```
Comment (dstrain115) [inline style.md:75]: nit: should be a comma after consistency.  Also, I would use stronger language, such as, "For the rules governing naming guidelines, see..."

File: cirq/circuits/circuit.py (hunk 7 above)
Comment (dabacon) [reply cirq/circuits/circuit.py:1500]: See nomenclature.md. For STATE_VECTOR_LIKE, state_rep or initial_state are ok, since it's not a vector (it can be an int).

File: cirq/qis/states_test.py (hunk 8 above)
Comment (dabacon) [reply cirq/qis/states_test.py:44]: I reworded these because these names are very non-standard.  Quantum computing folk should know |0>,|1>,|+>,|->,|+i>,|-i>.   Also they violated style.

File: cirq/qis/states_test.py (hunk 9 above)
Comment (dabacon) [reply cirq/qis/states_test.py:51]: typo. fixed.

File: cirq/sim/simulator.py (hunk 10 above)
Comment (dabacon) [reply cirq/sim/simulator.py:206]: Good call. fixed here and elsewhere.

File: cirq/sim/state_vector_simulator_test.py (hunk 11 above)
Comment (dabacon) [reply cirq/sim/state_vector_simulator_test.py:1]: This is a rename, think it should stay 2019?

File: docs/dev/nomenclature.md (hunk 12 above)
Comment (dabacon) [reply docs/dev/nomenclature.md:2]: Done

File: style.md (hunk 13 above)
Comment (dabacon) [reply style.md:75]: reworded

**Comment by dabacon** [comment]
Fixes #3022
